### ISSUE TITLE
certain upstart service on CentOS fails with enabled=yes

### ISSUE THREAD

**Issue by aabdnn** [ansible/ansible#7715]
##### Issue Type:

Bug Report
##### Ansible Version: 1.6.3

ansible 1.6.3
##### Environment:

From MacOS X 10.9.3, to CentOS 6.5.
##### Summary:

service module fails with enabled=yes
##### Steps To Reproduce:

Use the service module to ensure that an UPSTART job is running and enabled, like this:

ansible myhost -sK -m service -a 'name=riswhois state=started enabled=yes'
##### Expected Results:

I expect the service to be started if it were not running, and any related .override file in /etc/init to be removed or modified.
##### Actual Results:

```
myhost | FAILED >> {
"failed": true, 
"msg": "cannot detect command to enable service riswhois, typo or init system potentially unknown"
}
```

**Comment by jimi-c** [comment]
Scratch that, I didn't realize CentOS 6 had upstart capability. How did you install the service, via yum or manually?

**Comment by jimi-c** [comment]
And could you please share the upstart script for the service, especially if installed manually.

**Comment by aabdnn** [comment]
CentOS 6's init system _is_ upstart. But it still has SysV-style scripts for backward compatibility.

This specific upstart job is one I created by hand. It works fine on the command line, and I can start and stop the service. It also works with the service module of ansible. The only failure is when one passes the "enabled" flag. Then, ansible is supposed to create .override files.

My script is simple. Here's what it looks like:

```
start on stopped rc
stop on runlevel [016]
respawn
script
    [ -f /etc/sysconfig/riswhois ] && . /etc/sysconfig/riswhois || :
    exec /usr/sbin/riswhoisd $OPTIONS
end script
```

**Comment by jimi-c** [comment]
And the upstart enable/disable works from the CLI? If so, could you share the return code of that command?

**Comment by aabdnn** [comment]
On Tuesday, June 10, 2014, James Cammarata notifications@github.com wrote:

> And the upstart enable/disable works from the CLI? If so, could you share
> the return code of that command?
> 
> There is no command to enable/disable upstart scripts. Instead, you have to
> create an override file. So if the upstart script is called riswhois.conf
> then you create a matching riswhois.override containing stanzas to override
> startup. This used to work in earlier versions of ansible. The service
> module would create or remove .override files. Now it just fails.

Anand

**Comment by mpdehaan** [comment]
"There is no command to enable/disable upstart scripts"

It's initctl

Sounds like this does seem like there isn't enough of this service lying on the system for it to tell it's a upstart service.

There is various detection code in the service code that would explain this - do you have skills to help debug @arhbkb?  Look at ./hacking/test-module for an easy way to run the module free of ansible and see http://michaeldehaan.net/post/35403909347/tips-on-using-debuggers-with-ansible

**Comment by aabdnn** [comment]
Hi @mpdehaan. In an upstart-based system, initctl starts, stops and restarts services. It does NOT enable/disable them. As I wrote previously, disabling an upstart job is achieved by creating a service.override file, containing a stanza to override normal startup.

Next up, the service module DOES detect when a service is managed by upstart, by discovering that there is no /etc/init.d/service file, but instead a /etc/init/service.conf file. So the service module knows how to start, stop and restart a service under upstart.

As I have explained before, the failure is occurring when we pass the "enabled=yes" flag. It is then that something is going wrong with service type detection. I could look at this, but I'm on vacation, for a couple of weeks, and I don't really want to spend my vacation time looking at Python code. I've tried to provide enough notes to try and pinpoint the circumstances under which this failure occurs. If you guys haven't figured it out towards the end of June, then I will try to debug this when I come back from vacation.

**Comment by ebastos** [comment]
Would that help?
https://github.com/ebastos/ansible/commit/127f90f3d7be429365f724bcc3eb844ee1ebf157

**Comment by aabdnn** [comment]
Thanks @ebastos ! Your solution looks like it would fix the issue. In fact, for some reason I had not noticed your comment here, but I came back to this bug today, and was going to report that I had discovered why it was broken. @jimi-c any chance of incorporating this fix into a bugfix branch of 1.6?

I would also like to note that while browsing the code of the service module, I noticed that it inserts the word "manual" into a .override file to disable the service. However, this feature is new to upstart 0.6.7, whereas CentOS 6 has upstart 0.6.5. This means that on CentOS 6, service enable/disable will not actually work. Perhaps this can be added to the documentation of this module.

Nevertheless this patch will at least stop ansible from erroring out and stopping the playbook run altogether.

**Comment by jimi-c** [comment]
# Closing This Ticket

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next major release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
